src/lib.rs

```diff
@@ -0,0 +1,230 @@
+        let version_re = Regex::new(r"^(?P<major>\d+)\.(?P<minor>\d+)(?:\.(?P<patch>\d+))?(?:(?P<pre0>[a-z]+)(?P<pre1>\d*))?$").unwrap();
+        if let Some(captures) = version_re.captures(version_string) {
+            match captures.name("major")
+                .and_then(|x| u64::from_str(x.as_str()).ok()) {
+                    Some(x) => version.major = x,
+                    None => return Err("No major version number found")
+                }
+            match captures.name("minor")
+                .and_then(|x| u64::from_str(x.as_str()).ok()) {
+                    Some(x) => version.minor = x,
+                    None => return Err("No minor version number found")
+                }
+            match captures.name("patch")
+                .and_then(|x| u64::from_str(x.as_str()).ok()) {
+                    Some(x) => version.patch = x,
+                    None => {}
+                }
+            if let Some(pre_0) = captures.name("pre0").map(|x| x.as_str().to_string()) {
+                if captures.name("pre1").is_some() {
+                    if let Some(pre_1) = captures.name("pre1")
+                        .and_then(|x| u64::from_str(x.as_str()).ok()) {
+                            version.pre = Some((pre_0, pre_1))
+                        } else {
+                            return Err("Failed to convert prelease number to u64");
+                        }
+                } else {
+                    return Err("Failed to convert prelease number to u64");
+                }
+            }
+        } else {
+            return Err("Failed to parse input as version string")
+        }
+        Ok(version)
+    }
+
+    fn to_semver(&self) -> semver::Version {
+        // The way the semver crate handles prereleases isn't what we want here
+        // This should be fixed in the long term by implementing our own comparison
+        // operators, but for now just act as if prerelease metadata was missing, otherwise
+        // it is almost impossible to use this with nightly
+        semver::Version {
+            major: self.major,
+            minor: self.minor,
+            patch: self.patch,
+            pre: vec![],
+            build: vec![]
+        }
+    }
+
+    pub fn matches(&self, version_req: &str) -> Result<bool, semver::ReqParseError> {
+        let req = try!(semver::VersionReq::parse(version_req));
+        Ok(req.matches(&self.to_semver()))
+    }
+}
+
+impl Display for Version {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        match self.patch {
+            0 => try!(write!(f, "{}.{}", self.major, self.minor)),
+            _ => try!(write!(f, "{}.{}.{}", self.major, self.minor, self.patch))
+        }
+        if let Some(ref pre) = self.pre {
+            try!(write!(f, "{}{}", pre.0, pre.1));
+        };
+        Ok(())
+    }
+}
+
+pub fn firefox_version(binary: &Path) -> Result<AppVersion, &'static str>  {
+    let mut version: AppVersion = Default::default();
```

Comment:
Just `let mut version = AppVersion::new()` is cleaner here. It makes sense to use `Default::default()` as you do in its ctor.